File: tests/test_quad.py
from fractions import Fraction

from rat_trig.trigonom import quad


def test_quad() -> None:
    """Test quadrance"""
    # Test with integer vector
    vec_int = [3, 4]
    assert quad(vec_int) == 25

    # Test with float vector
    vec_float = [3.0, 4.0]
    assert quad(vec_float) == 25.0

    # Test with Fraction vector
    vec_frac = [Fraction(3, 5), Fraction(4, 5)]
    assert quad(vec_frac) == 1

    # Test with another integer vector
    vec_int2 = [1, 1]
    assert quad(vec_int2) == 2

    # Test with zero vector
    vec_zero = [0, 0]
    assert quad(vec_zero) == 0


if __name__ == "__main__":
    import pytest

    pytest.main()

File: cpp_ai/tests/test_quad.cpp
/**
 * @file test_quad.cpp
 * @brief Tests for quadrance function using doctest
 */

#include "doctest.h"
#include "rat_trig/trigonom.hpp"

TEST_CASE("Testing quad function") {
    SUBCASE("Test with integer vector") {
        std::array<int, 2> vector = {3, 4};
        CHECK(rat_trig::quad(vector) == 25);
    }

    SUBCASE("Test with float vector") {
        std::array<float, 2> vector = {3.0f, 4.0f};
        CHECK(rat_trig::quad(vector) == doctest::Approx(25.0f));
    }

    SUBCASE("Test with double vector") {
        std::array<double, 2> vector = {3.0, 4.0};
        CHECK(rat_trig::quad(vector) == doctest::Approx(25.0));
    }

    SUBCASE("Test with another integer vector") {
        std::array<int, 2> vector = {1, 1};
        CHECK(rat_trig::quad(vector) == 2);
    }

    SUBCASE("Test with zero vector") {
        std::array<int, 2> vector = {0, 0};
        CHECK(rat_trig::quad(vector) == 0);
    }

    SUBCASE("Test with fractions (using doubles)") {
        std::array<double, 2> vector = {3.0 / 5.0, 4.0 / 5.0};
        CHECK(rat_trig::quad(vector) == doctest::Approx(1.0).epsilon(1e-10));
    }
}
